File: projecteuler/0816/p816.py
from decimal import *


N = 2000000


def gen_points():
    points = []
    s = 290797
    for i in range(2 * N - 1):
        t = (s * s) % 50515093
        if i % 2 == 0:
            points.append((Decimal(s), Decimal(t)))
        s = t
    return points


def solve_naive(points):
    best = Decimal('Infinity')
    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            x = points[i][0] - points[j][0]
            y = points[i][1] - points[j][1]
            d2 = x * x + y * y
            best = min(best, d2.sqrt())
    return best, sorted(points, key=lambda t: t[1])


def merge(a, b):
    c = []
    i, j = 0, 0
    while i < len(a) and j < len(b):
        if a[i][1] <= b[j][1]:
            c.append(a[i])
            i += 1
        else:
            c.append(b[j])
            j += 1
    while i < len(a):
        c.append(a[i])
        i += 1
    while j < len(b):
        c.append(b[j])
        j += 1
    return c


def closest_pair_distance(points):
    if len(points) <= 5:
        return solve_naive(points)
    mid = len(points) // 2
    left, right = points[:mid], points[mid:]
    l = (left[-1][0] + right[0][0]) / 2
    d1, left = closest_pair_distance(left)
    d2, right = closest_pair_distance(right)
    d, merged = min(d1, d2), merge(left, right)
    remaining = [(x, y) for x, y in merged \
                 if abs(l - x) <= d]
    for i, (x, y) in enumerate(remaining):
        for j in range(i - 3, i + 4):
            if j >= 0 and j < len(remaining) and j != i:
                d2 = (x - remaining[j][0]) ** 2 + (y - remaining[j][1]) ** 2
                d = min(d, d2.sqrt())
    return d, merged


def solve(points):
    points = sorted(points, key=lambda t: t[0])
    return closest_pair_distance(points)


points = gen_points()
#print('{:.9f}'.format(solve_naive(points)))
d, _ = solve(points)
print('{:.9f}'.format(d))

File: projecteuler/0816/p816.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
using namespace std;

const int N = 2000000;

vector< pair<int,int> > gen_points() {
    vector< pair<int,int> > points;
    long long s = 290797, t = 0;
    for (int i = 0; i < 2 * N - 1; i++) {
        t = (s * s) % 50515093;
        if (!(i & 1)) points.push_back(make_pair(s, t));
        s = t;
    }
    return points;
}

double solve_naive(vector< pair<int,int> > &points) {
    double best = 1e9;
    for (int i = 0; i < points.size(); i++)
        for (int j = i + 1; j < points.size(); j++) {
            double x = points[i].first - points[j].first;
            double y = points[i].second - points[j].second;
            best = min(best, sqrt(x * x + y * y));
        }
    sort(points.begin(), points.end(), [](auto &a, auto &b) {
        return  a.second < b.second;
    });
    return best;
}

vector< pair<int,int> > merge(vector< pair<int,int> > &a, vector< pair<int,int> > &b) {
    vector< pair<int,int> > c;
    int i = 0, j = 0;
    while (i < a.size() && j < b.size()) {
        if (a[i].second < b[j].second) {
            c.push_back(a[i++]);
        } else {
            c.push_back(b[j++]);
        }
    }
    while (i < a.size()) c.push_back(a[i++]);
    while (j < b.size()) c.push_back(b[j++]);
    return c;
}

double closest_pair_distance(vector< pair<int,int> > &points) {
    if (points.size() <= 5) return solve_naive(points);
    int m = points.size() >> 1;
    auto left = vector< pair<int,int> >(points.begin(), points.begin() + m);
    auto right = vector< pair<int,int> >(points.begin() + m , points.end());
    double l = (right[0].first + left[left.size() - 1].first) * 0.5;
    double d1 = closest_pair_distance(left);
    double d2 = closest_pair_distance(right);
    double d = min(d1, d2);
    auto merged = merge(left, right);
    vector< pair<int,int> > remaining;
    for (auto p : merged) {
        if (abs(p.first - l) < d) remaining.push_back(p);
    }
    for (int i = 0; i < remaining.size(); i++) {
        for (int j = i - 3; j < i + 4; j++) {
            if (j >= 0 && j < remaining.size() && j != i) {
                double x = remaining[i].first - remaining[j].first;
                double y = remaining[i].second - remaining[j].second;
                d = min(d, sqrt(x * x + y * y));
            }
        }
    }
    return d;
}

int main() {
    vector< pair<int,int> > points = gen_points();
    sort(points.begin(), points.end(), [](auto &a, auto &b) {
        return a.first < b.first;
    });
    double d = closest_pair_distance(points);
    cout << fixed << setprecision(9) << d << '\n';
    return 0;
}
